### PR TITLE
[clang] Fix underlying type of EmbedExpr

### PR DESCRIPTION
**Pull request by Fznamznon**
This patch makes remaining cases of #embed to emit int type since there is an agreement to do that for C. C++ is being discussed, but in general we don't want to produce different types for C and C++.

### CODE REVIEW

File: clang/test/Preprocessor/embed_codegen.cpp [hunk 1]
```diff
@@ -13,9 +13,9 @@ int ca[] = {
 };
 
 // CHECK: %arrayinit.element = getelementptr inbounds i32, ptr %notca, i64 1
-// CHECK: store i8 106, ptr %arrayinit.element, align 4
+// CHECK: store i32 106, ptr %arrayinit.element, align 4
```
[the remark this answers is not in the document]
Comment (cor3ntin) [reply clang/test/Preprocessor/embed_codegen.cpp:16]: I was trying to write a test case and found an unrelated embed crash https://compiler-explorer.com/z/bMWbGfh79 

can we add the test here https://github.com/llvm/llvm-project/pull/97274#issuecomment-2230410564 (with sizeof)

**Review by AaronBallman — COMMENTED** [review]
Changes LGTM but I agree with @cor3ntin that we should add the test case from https://github.com/llvm/llvm-project/pull/97274#issuecomment-2230410564 explicitly. Some other tests to consider adding at the same time:
```
static_assert(_Generic(
#embed __FILE__ limit(1)
  , int : 1, default : 0));

static_assert(alignof(typeof(
#embed __FILE__ limit(1)
)) == alignof(int));

printf("%hhu", // Do we get a -Wformat diagnostic about the specifier mismatch?
#embed __FILE__ limit(1)
);
```

**Comment by Fznamznon** [comment]
I can add test with sizeof for sure, the problem is that it is already working https://godbolt.org/z/K69v8KPYa and we have similar ones in codebase already.
The test I'm struggling with is to check that elements in initializer list that are expanded from embed also have int type.

**Comment by AaronBallman** [comment]
> I can add test with sizeof for sure, the problem is that it is already working https://godbolt.org/z/K69v8KPYa and we have similar ones in codebase already. 

Ah, if we have sufficient coverage already, that's fine. I was just thinking about coverage in general, didn't check to see if it was redundant. :-)

> The test I'm struggling with is to check that elements in initializer list that are expanded from embed also have int type.

If you have an embed file with a single byte in it whose value is 0xFF, you could use a test like:
```
struct S {
  signed char ch;
};

constexpr struct S s = {
#embed "foo"
};
```
which should give the same behavior as: https://godbolt.org/z/Tno5YPexM

**Comment by Fznamznon** [comment]
> If you have an embed file with a single byte in it whose value is 0xFF, you could use a test like:
struct S {
  signed char ch;
};
constexpr struct S s = {
#embed "foo"
};
which should give the same behavior as: https://godbolt.org/z/Tno5YPexM

Thanks! That worked.

I didn't add these though
```
static_assert(_Generic(
#embed __FILE__ limit(1)
  , int : 1, default : 0));

static_assert(alignof(typeof(
#embed __FILE__ limit(1)
)) == alignof(int));

printf("%hhu", // Do we get a -Wformat diagnostic about the specifier mismatch?
#embed __FILE__ limit(1)
);
```
for several reasons:

- They don't test this patch as it affects only what happens in an initializer list
- The one with alignof/typeof asserts for C++ (not for C) and I like to make a fix and add this test in a separate PR
- we don't emit warnings even for `printf("%hhu", 300);`

**Comment by Fznamznon** [comment]
> The one with alignof/typeof asserts for C++ (not for C) and I'd like to make a fix and add this test in a separate PR

Posted https://github.com/llvm/llvm-project/pull/99624 to fix the assertion

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**